book/internals/hooks/after_build_hook.nim

```diff
@@ -0,0 +1,46 @@
+import nimib, nimibook
+
+nbInit(theme = useNimibook)
+nbText: """
+## **AfterBuild Hook**
+The `afterBuild`-hook runs once after initial values (default-values, values passed in by other widgets during instantiation) have been assigned to the `WidgetState`.
+
+They are useful if any processing on the initial data that is passed in must happen. Example are validating data, inferring data from passed in data, or fetching other data based on what was passed in. In renderables they are also useful to update the GTK widget once with data from the initial `WidgetState`.
+
+It should be noted that unlike `build` hooks, `afterBuild` hooks are not inherited by any child-widget. For more information, see the `build` hooks section.
+
+Let's return to our earlier renderable example and write it so that the parent-widget can decide what text the `MyRenderable` should display: 
+"""
+
+nbCode:
+  import owlkettle
+  import owlkettle/gtk
+  import std/json
+
+  renderable MyRenderable:
+    text: string
+      
+    hooks:
+      beforeBuild:
+        echo state.repr
+        state.internalWidget = gtk_label_new("")
+      
+      afterBuild:
+        gtk_label_set_text(state.internalWidget, state.text.cstring)
+
+  viewable App:
+    discard
+
+  method view(app: AppState): Widget =
+    result = gui:
+      Window:
+        MyRenderable(text = "Defined by App")
+
+  # brew(gui(App())) # Uncomment to execute app
```

Comment:
Should not be commented out in the docs. Maybe we can add a flag like `-d:owlkettleDoNotBrew` to just directly `return` from `brew`

Reply:
Instinctively my opinion is that only code that adds/fixes features and contributes to owlkettle doing its actual task should be part of the actual source. The actual source should never have to change for things outside of it , such as tests or documentation, as the necessity for such a feature is non-obvious.

Is the issue that you would want to get more guarantees that certain things work by having brew run, execute most of its code and return *just* before it would launch a GUI?

Reply:
This is not about guarantees of correctness of the examples, but rather about the readability of the docs. I think you should be able to just copy/paste the examples from the docs without any modification. As a reader of the docs I would ask myself "Why do I have to uncomment this before I can try it out?"

From a software engineering perspective, I think we are long past the point of cleanly separating docs & actual implementation; I mean there is basically a whole documentation generator hidden behind the `-d:owlkettleDocs` flag, so adding another flag would not really hurt. :smile:

Reply:
If owlkettleDocs is already a thing, mind if we use that very same switch? I see no real benefit in having them separate when they're both for docs in their own way. I'd maybe apply a comment explaining as to why a "when defined(owlkettleDocs): return` is necessary in brew as well

Reply:
I just skimmed through the brew proc implementations and noticed that we have 3 `brew` procs and their implementations are all different from one another. 

The first generates an `AppConfig` instance and then just runs `runMainLoop`.
The second generates an `AppConfig` instance, an `activateCallback` and then instantiates an application and calls seemingly completely different functions, even though superficially it is identical to the other one with only the "id"-parameter differing
The third is seemingly like the first but with provides a style-manager.

Thus I see no super easy way to implement this cleanly in a single place.
I think putting a `when defined/owlkettleDocs: return` switch could be added to `runMainLoop` ? But that still leaves out the second `brew` implementation which doesn't use `runMainLoop`.  What's the way to go here?

Reply:
@can-lehmann I've decided to just go with applying the following constructs to all owlkettle examples:

```nim
when not defined(owlkettleDocs):
  brew(gui(App()))
``` 

That should do